src/target/riscv/riscv.c

```diff
@@ -1561,6 +1561,100 @@ int riscv_remove_watchpoint(struct target *target,
+	if (tdata1 & hit_mask)
+		return M_HIT;
+	const riscv_reg_t tdata1_test = set_field(tdata1, CSR_MCONTROL_HIT, 1);
+	if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test) != ERROR_OK)
+		return M_HIT_ERROR;
+	riscv_reg_t tdata1_test_rb;
+	if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+		return M_HIT_ERROR;
+	int tdata1_test_hit = get_field(tdata1_test, CSR_MCONTROL_HIT);
+	int tdata1_test_rb_hit = get_field(tdata1_test_rb, CSR_MCONTROL_HIT);
+	if (tdata1_test_rb_hit != tdata1_test_hit)
+		return M_HIT_NOT_SUPPORTED;
+
+	if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+		return M_HIT_ERROR;
+	return M_NOT_HIT;
+}
+
+typedef enum {
+	M6_HIT_ERROR,
+	M6_HIT_NOT_SUPPORTED,
+	M6_NOT_HIT,
+	M6_HIT_BEFORE,
+	M6_HIT_AFTER,
+	M6_HIT_IMM_AFTER
+} mctrl6hitstatus;
+
+static mctrl6hitstatus check_mcontrol6_hit_status(struct target *target,
+		riscv_reg_t tdata1, uint64_t hit_mask)
+{
+	int hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+	int hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+	int hit_info = (hit1 << 1) | hit0;
+	if (hit_info == CSR_MCONTROL6_HIT0_BEFORE)
+		return M6_HIT_BEFORE;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_AFTER)
+		return M6_HIT_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_IMMEDIATELY_AFTER)
+		return M6_HIT_IMM_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_FALSE) {
+		/* Check timing "before" support. */
+		riscv_reg_t tdata1_hit_before_test = set_field(tdata1, CSR_MCONTROL6_HIT0, 1);
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_before_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		riscv_reg_t tdata1_test_rb;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_before_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
+
+		/* Check timing "after" support. */
+		riscv_reg_t tdata1_hit_after_test = set_field(tdata1, CSR_MCONTROL6_HIT1, 1);
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_after_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_after_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
+
+		/* Check timing "imm after" support. */
+		riscv_reg_t tdata1_hit_imm_after_test = tdata1_hit_before_test | tdata1_hit_after_test;
+		if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_imm_after_test) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+			return M6_HIT_ERROR;
+		if (tdata1_test_rb == tdata1_hit_imm_after_test) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			return M6_NOT_HIT;
+		}
```

Comment:
Please, consider looping over an array of `tdata1` values to check:
```suggestion
	if (hit_info == CSR_MCONTROL6_HIT0_FALSE) {
		riscv_reg_t tdata1_tests[] = {
			set_field(tdata1, CSR_MCONTROL6_HIT0, 1),
			set_field(tdata1, CSR_MCONTROL6_HIT1, 1),
			set_field(tdata1, CSR_MCONTROL6_HIT0, 1) | field_value(CSR_MCONTROL6_HIT0, 1)
		};
		for (int i = 0; i < ARRAY_SIZE(tdata1_tests); ++i) {
			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_hit_before_test) != ERROR_OK)
				return M6_HIT_ERROR;
			riscv_reg_t tdata1_test_rb;
			if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
				return M6_HIT_ERROR;
			if (tdata1_test_rb == tdata1_hit_before_test) {
				if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
					return M6_HIT_ERROR;
				return M6_NOT_HIT;
			}
		}
```